src/downloader.rs

```diff
@@ -0,0 +1,236 @@
+use reqwest::{Client, Url};
+use std::{path::PathBuf, sync::Arc};
+use tokio::{
+    fs::File,
+    io::AsyncWriteExt,
+    sync::{mpsc, oneshot, watch, Semaphore},
+};
+
+use crate::Error;
+
+const QUEUE_SIZE: usize = 100;
+const MAX_RETRIES: usize = 3;
+
+#[derive(Debug)]
+struct DownloadRequest {
+    url: Url,
+    destination: PathBuf,
+    result: oneshot::Sender<Result<File, Error>>,
+    status: watch::Sender<Status>,
+    cancel: oneshot::Receiver<()>,
+    remaining_retries: usize,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct DownloadProgress {
+    pub bytes_downloaded: u64,
+    pub total_bytes: Option<u64>,
+}
+
+#[derive(Debug)]
+pub struct DownloadHandle {
+    result: oneshot::Receiver<Result<File, Error>>,
+    status: watch::Receiver<Status>,
+    cancel: oneshot::Sender<()>,
+}
+
+impl std::future::Future for DownloadHandle {
+    type Output = Result<tokio::fs::File, Error>;
+
+    fn poll(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        use std::pin::Pin;
+        use std::task::Poll;
+        match Pin::new(&mut self.result).poll(cx) {
+            Poll::Ready(Ok(result)) => Poll::Ready(result),
+            Poll::Ready(Err(e)) => Poll::Ready(Err(Error::Oneshot(e))),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl DownloadHandle {
+    pub fn status(&self) -> Status {
+        self.status.borrow().clone()
+    }
+
+    pub fn cancel(self) {
+        self.cancel.send(()).ok();
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Status {
+    Pending,
+    InProgress(DownloadProgress),
+    Completed,
+    Retrying,
+    Cancelled,
+    Failed,
+}
+
+#[derive(Debug)]
+pub struct DownloadManager {
+    queue: mpsc::Sender<DownloadRequest>,
+    semaphore: Arc<Semaphore>,
+}
+
+impl Drop for DownloadManager {
+    fn drop(&mut self) {
+        // Need to manually close the semaphore to make sure dispatcher_thread stops waiting for permits
+        self.semaphore.close();
+    }
+}
+
+impl DownloadManager {
+    pub fn new(limit: usize) -> Self {
+        let (tx, rx) = mpsc::channel(QUEUE_SIZE);
+        let client = Client::new();
+        let semaphore = Arc::new(Semaphore::new(limit));
+        let manager = Self {
+            queue: tx,
+            semaphore: semaphore.clone(),
+        };
+        // Spawn the dispatcher thread to handle download requests
+        tokio::spawn(async move { dispatcher_thread(client, rx, semaphore).await });
+        manager
+    }
+
+    pub fn set_max_parallel_downloads(&self, limit: usize) {
+        let current = self.semaphore.available_permits();
+        if limit > current {
+            self.semaphore.add_permits(limit - current);
+        } else if limit < current {
+            let to_remove = current - limit;
+            for _ in 0..to_remove {
+                let _ = self.semaphore.try_acquire();
+            }
+        }
+    }
```

Comment:
On looking carefully, This has a race condition so this needs to be reimplemented.